sparql/kidney-three-hop-neighborhood-down.rq

```diff
@@ -0,0 +1,63 @@
+PREFIX part: <http://purl.org/vocab/participation/schema#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX connected_to: <http://purl.obolibrary.org/obo/RO_0002170>
+PREFIX part_of: <http://purl.obolibrary.org/obo/BFO_0000050>
+PREFIX hint: <http://www.bigdata.com/queryHints#>
+# specify focus node here
+PREFIX node: <http://purl.obolibrary.org/obo/UBERON_0002113> 
+SELECT ?s (MIN(?s_label) AS ?slabel) ?p (MIN(?p_label) AS ?plabel) ?o (MIN(?o_label) AS ?olabel)
+WHERE {
+  {
+    SELECT DISTINCT ?s
+    WHERE {
+      GRAPH <http://reasoner.renci.org/nonredundant> {
+        VALUES ?o { node: }
+        VALUES ?p_p { connected_to: part_of: }
+        ?s ?p ?o .
+        ?p rdfs:subPropertyOf* ?p_p .
+      }
+    }
+  }
+  UNION
+  {
+    SELECT DISTINCT ?s
+    WHERE {
+      GRAPH <http://reasoner.renci.org/nonredundant> {
+        VALUES ?o { node: }
+        VALUES ?pn_p { connected_to: part_of: }
+        ?s ?p1 ?o1 .
+        ?o1 ?p2 ?o .
+        ?p1 rdfs:subPropertyOf* ?pn_p .
+        ?p2 rdfs:subPropertyOf* ?pn_p .
+      }
+    }
+  }
+  UNION
+  {
+    SELECT DISTINCT ?s
+    WHERE {
+      GRAPH <http://reasoner.renci.org/nonredundant> {
+        VALUES ?o { node: }
+        VALUES ?pn_p { connected_to: part_of: }
+        ?s ?p1 ?o1 .
+        ?o1 ?p2 ?o2 .
+        ?o2 ?p3 ?o .
+        ?p1 rdfs:subPropertyOf* ?pn_p .
+        ?p2 rdfs:subPropertyOf* ?pn_p .
+        ?p3 rdfs:subPropertyOf* ?pn_p .
+      }
+    }
+  }
+  UNION { VALUES ?o { node: } }
+  # this hint is required for Blazegraph to run this query efficiently
+  hint:Prior hint:runFirst true .
+  GRAPH <http://reasoner.renci.org/nonredundant> {
+      VALUES ?p_p { connected_to: part_of: }
+      ?s ?p ?o .
+      ?p rdfs:subPropertyOf* ?p_p .
+  }
```

Comment:
@balhoff, how could I also get the `rdfs:subClassOf` here? I tried using UNION, but I'm getting QueryTimeoutException after 1 minute.

Reply:
Maybe missing from nonredundant graph at present?  Could try 

<http://reasoner.renci.org/ontology>

Reply:
I believe it's in there. I'll play with this query a bit.

Reply:
I think the issue you're having with `rdfs:subClassOf` being missing is that it has no label. Here is a version with subclass; it returns for me in 16.5 seconds:

```sparql
PREFIX part: <http://purl.org/vocab/participation/schema#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX connected_to: <http://purl.obolibrary.org/obo/RO_0002170>
PREFIX part_of: <http://purl.obolibrary.org/obo/BFO_0000050>
PREFIX hint: <http://www.bigdata.com/queryHints#>
# specify focus node here
PREFIX node: <http://purl.obolibrary.org/obo/UBERON_0002113> 
SELECT ?s (MIN(?s_label) AS ?slabel) ?p (MIN(?p_label) AS ?plabel) ?o (MIN(?o_label) AS ?olabel)
WHERE {
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        VALUES ?o { node: }
        VALUES ?p_p { connected_to: part_of: rdfs:subClassOf }
        ?s ?p ?o .
        ?p rdfs:subPropertyOf* ?p_p .
      }
    }
  }
  UNION
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        VALUES ?o { node: }
        VALUES ?pn_p { connected_to: part_of: rdfs:subClassOf }
        ?s ?p1 ?o1 .
        ?o1 ?p2 ?o .
        ?p1 rdfs:subPropertyOf* ?pn_p .
        ?p2 rdfs:subPropertyOf* ?pn_p .
      }
    }
  }
  UNION
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        VALUES ?o { node: }
        VALUES ?pn_p { connected_to: part_of: rdfs:subClassOf }
        ?s ?p1 ?o1 .
        ?o1 ?p2 ?o2 .
        ?o2 ?p3 ?o .
        ?p1 rdfs:subPropertyOf* ?pn_p .
        ?p2 rdfs:subPropertyOf* ?pn_p .
        ?p3 rdfs:subPropertyOf* ?pn_p .
      }
    }
  }
  UNION { VALUES ?o { node: } }
  # this hint is required for Blazegraph to run this query efficiently
  hint:Prior hint:runFirst true .
  GRAPH <http://reasoner.renci.org/nonredundant> {
      VALUES ?p_p { connected_to: part_of: rdfs:subClassOf }
      ?s ?p ?o .
      ?p rdfs:subPropertyOf* ?p_p .
  }
  ?s rdfs:label ?s_label .
  { ?p rdfs:label ?p_label . } UNION { VALUES (?p ?p_label) { (rdfs:subClassOf "rdfs:subClassOf") } }
  ?o rdfs:label ?o_label .
}
GROUP BY ?s ?p ?o
```

Reply:
Thank you, @balhoff!

Reply:
I think there's another sneaky issue with the current version. Here, `?p1` and `?p2` are forced to be subproperties from the same property in your list, rather than any of them:

```sparql
VALUES ?pn_p { connected_to: part_of: rdfs:subClassOf }
?s ?p1 ?o1 .
?o1 ?p2 ?o .
?p1 rdfs:subPropertyOf* ?pn_p .
?p2 rdfs:subPropertyOf* ?pn_p .
```

This version returns more results:

```sparql
PREFIX part: <http://purl.org/vocab/participation/schema#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX connected_to: <http://purl.obolibrary.org/obo/RO_0002170>
PREFIX part_of: <http://purl.obolibrary.org/obo/BFO_0000050>
PREFIX hint: <http://www.bigdata.com/queryHints#>
# specify focus node here
PREFIX node: <http://purl.obolibrary.org/obo/UBERON_0002113> 
SELECT ?s (MIN(?s_label) AS ?slabel) ?p (MIN(?p_label) AS ?plabel) ?o (MIN(?o_label) AS ?olabel)
WHERE {
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        {
          SELECT DISTINCT ?p WHERE {
            VALUES ?p_p { connected_to: part_of: rdfs:subClassOf }
            GRAPH <http://reasoner.renci.org/ontology> {
              ?p rdfs:subPropertyOf* ?p_p .
            }
          }
        }
        VALUES ?o { node: }
        ?s ?p ?o .
      }
    }
  }
  UNION
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        {
          SELECT DISTINCT ?p1 ?p2 WHERE {
            VALUES ?p1_p { connected_to: part_of: rdfs:subClassOf }
            VALUES ?p2_p { connected_to: part_of: rdfs:subClassOf }
            GRAPH <http://reasoner.renci.org/ontology> {
              ?p1 rdfs:subPropertyOf* ?p1_p .
              ?p2 rdfs:subPropertyOf* ?p2_p .
            }
          }
        }
        VALUES ?o { node: }
        ?s ?p1 ?o1 .
        ?o1 ?p2 ?o .
      }
    }
  }
  UNION
  {
    SELECT DISTINCT ?s
    WHERE {
      GRAPH <http://reasoner.renci.org/nonredundant> {
        {
          SELECT DISTINCT ?p1 ?p2 ?p3 WHERE {
            VALUES ?p1_p { connected_to: part_of: rdfs:subClassOf }
            VALUES ?p2_p { connected_to: part_of: rdfs:subClassOf }
            VALUES ?p3_p { connected_to: part_of: rdfs:subClassOf }
            GRAPH <http://reasoner.renci.org/ontology> {
              ?p1 rdfs:subPropertyOf* ?p1_p .
              ?p2 rdfs:subPropertyOf* ?p2_p .
              ?p3 rdfs:subPropertyOf* ?p3_p .
            }
          }
        }
        VALUES ?o { node: }
        ?s ?p1 ?o1 .
        ?o1 ?p2 ?o2 .
        ?o2 ?p3 ?o .
      }
    }
  }
  UNION { VALUES ?o { node: } }
  # this hint is required for Blazegraph to run this query efficiently
  hint:Prior hint:runFirst true .
  GRAPH <http://reasoner.renci.org/nonredundant> {
    {
          SELECT DISTINCT ?p WHERE {
            VALUES ?p_p { connected_to: part_of: rdfs:subClassOf }
            GRAPH <http://reasoner.renci.org/ontology> {
              ?p rdfs:subPropertyOf* ?p_p .
            }
          }
        }
      ?s ?p ?o .
  }
  ?s rdfs:label ?s_label .
  { ?p rdfs:label ?p_label . } UNION { VALUES (?p ?p_label) { (rdfs:subClassOf "rdfs:subClassOf") } }
  ?o rdfs:label ?o_label .
}
GROUP BY ?s ?p ?o
```

I think there's a cleaner way to do it using FILTER; just posting that version as a first pass.

Reply:
Here's the graph visualization of this last query. 
What do you think, @dosumis?

![kidney-three-hop-neighborhood-down](https://user-images.githubusercontent.com/2208124/141323301-2b481100-eda6-40ec-a2cd-5d0e05c34ef2.png)

Reply:
@anitacaron this version returns even more results (627). I expected them to have the same answers, so I must be making a mistake in at least one of the queries. What do you think? https://api.triplydb.com/s/dtfkSHMVP

Reply:
@balhoff - your one returns subClassOf edges, which are not in the graph Anita posted.

Reply:
It's a consequence of defining all `?s` as those reachable by 1 to 3 hops from the term. Then getting any edges with those subjects and the preferred relations. An alternative approach would be to get all those `?s` and then query for only edges between any of them (to do that you have to duplicate some code to get them "twice"). I'll try another version that shows that.

Reply:
@anitacaron @dosumis please test visualization of this one: https://api.triplydb.com/s/s5avRLI6w  Have to do some Blazegraph tricks to not time out.

Reply:
![kidney-three-hop-neighborhood-down-inside](https://user-images.githubusercontent.com/2208124/141337012-3bc6ac53-dc55-439e-aba6-fc602c9528c6.png)

Reply:
Perfect!

Clearly too much for figure for paper, but we can use the same code to look at graph for some organ part (e.g. nephron) and compare this to the same in CCF.owl)